src/armature/Animation.ts

```diff
@@ -0,0 +1,176 @@
+        while (queue.length > 0 && queue[0].start <= currentTime) {
+            current.push(queue[0]);
+            queue.shift();
+        }
+    }
+
+    /**
+     * Updates the nodes for all currently active animations, interpolating between the given
+     * animation states.
+     */
+    function applyCurrentAnimations(currentTime: number) {
+        // TODO: use curves that aren't linear
+        current.forEach((animation: AnimationDescription) => {
+            const currentTransform = animation.node.getRawTransformation();
+            const period = animation.duration + animation.repeatDelay;
+
+            let currentCycle = Math.floor((currentTime - animation.start) / period);
+            if (animation.times > 0) {
+                currentCycle = Math.min(animation.times - 1, currentCycle);
+            }
+
+            const timeInCurrentCycle = currentTime - animation.start - currentCycle * period;
+            const remainingTargetsExist = animation.times === 0 || currentCycle < animation.times;
+            const inNewCycle = animation.lastCycle < currentCycle;
+
+            // Recompute target state of animation if we have entered a new cycle
+            if (animation.finalTransform === null || (inNewCycle && remainingTargetsExist)) {
+                animation.lastTimeInCycle = 0;
+
+                // Clone the current node and apply the callback to generate the target state
+                const finalNode = Node.clone(animation.node);
+                animation.to(finalNode);
+
+                // Extract the transformation from the modified node
+                animation.finalTransform = finalNode.getRawTransformation();
+            }
+
+            const timeRemainingInCycle = animation.duration - animation.lastTimeInCycle;
+            let amount = (timeInCurrentCycle - animation.lastTimeInCycle) / timeRemainingInCycle;
+
+            // If we are in the delay period between cycles, do nothing and exit early
+            if (amount > 1 && remainingTargetsExist) {
+                return;
+            }
+
+            amount = Math.max(0, amount);
+            amount = Math.min(1, amount);
+
+            const interpolated = currentTransform.interpolate(animation.finalTransform, amount);
+            animation.node.setRawTransformation(interpolated);
+
+            animation.lastTimeInCycle = timeInCurrentCycle;
+            animation.lastCycle = currentCycle;
+        });
+    }
+
+    /**
+     * Remove animations that have entirely completed from `current`.
+     */
+    function removeFinishedAnimations(currentTime: number) {
+        remove(current, (animation: AnimationDescription) => {
+            if (animation.times === 0) {
+                return false;
+            } else if (animation.times === 1) {
+                return (currentTime - animation.start) / animation.duration >= 1;
+            } else {
+                return (
+                    (currentTime - animation.start) /
```

Comment:
optional nit: reading this line was a bit tough because it's split over 3 lines. may be easier to break down the numerator and denominator into variables like `totalAnimationTime` and `animationCycleTime` or something?